framework/data/BaseDataProvider.php

```diff
@@ -44,6 +50,20 @@
     private $_keys;
     private $_models;
     private $_totalCount;
+    
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        parent::init();
+        if (!isset($this->id)) {
+            if (self::$counter > 0) {
```

Comment:
Could be done with a trait: http://php.net/manual/en/language.oop5.traits.php#language.oop5.traits.static

Doing it with a class would open implementation details via public API.